_posts/2018-07-17-its-a-lot-more-snappy.md

```diff
@@ -0,0 +1,171 @@
+                await this.table.bulkPut(partition);
+                partitionStart = partitionEnd;
+            }
+        });
+    }
+}
+```
+
+Now that we had implemented this pattern, we noticed that the initial download of the data was taking much longer than it had previously. We were still downloading the dataset in the same way, performing a mapping operation on the dataset before storing it in our redux store and then inserting all of our newly obtained Pokemon into our Pokemon DB. If only there existed a method, which could enable us to parallelise these operations.
+
+It was at this point our fearless tech lead, William Sia(TM) proposed using a web worker for that exact purpose. This was new and very exciting. Of course we wanted to use a web worker!
+
+I found a module called [worker-loader](https://www.npmjs.com/package/worker-loader) which seamlessly integrated with our current build and packaging eco-system. In fact, worker-loader almost made it inconceivable that we waited such a long time to implement our first web worker. We didn’t even need to add any additional config to our web pack config files.
+
+When using a web worker, all that's needed to get things started is:
+
+* initialise the worker
+* post a message to the worker (preferably as a string)
+* create a callback for when the worker completes
+
+_from the main thread:_
+
+```typescript
+function spawnWorker(): void {
+    const bulkPutWorker = new Worker();
+    bulkPutWorker.postMessage(JSON.stringify({ pokemon: this.pokemon, location: this.location }));
+    pokemonActions.startInsertingPokemon();
+    window.addEventListener('pokemonUpdated', pokemonActions.endInsertingPokemon);
+    bulkPutWorker.onmessage = (event: MessageEvent): void => {
+        window.dispatchEvent(new Event(event.data));
+    };
+}
+```
+
+_within the worker:_
+
+```typescript
+worker.onmessage = async (event: MessageEvent): Promise<void> => {
+
+    const { location, pokemon } = JSON.parse(event.data) as BulkPutEvent;
+    const pokemonUpdater = new PokemonUpdater(location, pokemon);
+    try {
+        await pokemonUpdater.updatePokemon();
+    } catch (error) {
+        console.log('Something has gone awry', error);
+    }
+    worker.postMessage('pokemonUpdated');
+    self.close();
+};
+```
+
+Adopting this approach meant that we offload the computationally expensive task of writing all this data to IndexedDB. The user can now continue on the main thread and the app will remain responsive without any noticeable decrease in responsiveness over time. Having all this data stored in IndexedDB also means, offline usage is now possible and subsequent loads of the dataset are no longer a concern at all because the data can be accessed instantly from the local store.
+
+The Chrome dev tools also reveal that using a background process does not result in a significant increase in memory usage. Having noticed these improvements, our PM said “After testing the app, it’s now a lot more snappy!”
```

Comment:
Feels like there is a conclusion missing here.

Part of that - PWA itself wasn't the problem?  It was our implementation/the data we were dealing with.  Using PWA gave us tools to help solve the problem.  (Is that an accurate assessment?)

Reply:
Without the conclusion, the post falls a little flat.

I like `Having noticed these improvements, our PM said “After testing the app, it’s now a lot more snappy!”` as the last line.  

At the moment it's lost.  Also `client` or `customer` rather than `PM`? - sounds better.